site/topics/lists/lists.rst

```diff
@@ -112,13 +112,173 @@ Mutability
+* The above functionality can be achieved with a ``while`` loop, but the code needed is a little more cumbersome
+
+.. code-block:: python
+    :linenos:
+
+    i = 0
+    while i < 5:
+      print(i)
+      i += 1
+
+    # Results in
+    #   0
+    #   1
+    #   2
+    #   3
+    #   4
+
+
+.. admonition:: Activity
+    :class: activity
+
+    Write a function ``beer_on_wall`` that prints out ``"n bottles of beer on the wall"`` for all ``n`` from ``99`` down
+    to ``0``. This function must use a ``for`` loop with the ``range`` function.
+
+    The difficulty here is the need to count backwards.
+
+    `Perhaps a read of the documentation can help. <https://docs.python.org/3/library/stdtypes.html#typesseq-range>`_
+
+    .. raw:: html
+
+        <iframe width="560" height="315" src="https://www.youtube.com/embed/0AuMKIMiPMo" frameborder="0" allowfullscreen></iframe>
+
+
+Looping Over Indices and Enumerate
+----------------------------------
+
+* Sometimes it is important to know the indices of each element being printed out
+* For example, if the goal is to do a linear search to find the index of a given element, the following would be problematic
+
+.. code-block:: python
+    :linenos:
+
+    def index_of_uhoh(needle, haystack):
+        for element in haystack:
+            if element == needle:
+                return ???? # How do I find the index?
+        return -1
+
+
+* Using the ``range`` function would help in this situation as we can use it to loop over each index in ``haystack``
+* The only catch is that we need to actually index ``haystack``
+
+.. code-block:: python
+    :linenos:
+
+    def index_of(needle, haystack):
+        haystack_length = len(haystack)
+        for i in range(haystack_length):
+            if haystack[i] == needle:
+                return i
+        return -1
+
+
+* To make what is going on a little clearer, consider the following example
+
+.. code-block:: python
+    :linenos:
+
+    another_list = ["a", "b", "c", "d", "e"]
+    for i in range(len(another_list)):
+        print(i, another_list[i])
+
+    # Results in
+    #   0 a
+    #   1 b
+    #   2 c
+    #   3 d
+    #   4 e
+
+
+* This loop prints out the index (``i``) along with the element at the given index (``another_list[i]``)
+* This functionality is quite common and Python even provides a shorthand for achieving it --- ``enumerate``
+
+.. code-block:: python
+    :linenos:
+
+    another_list = ["a", "b", "c", "d", "e"]
+    for i, element in enumerate(another_list):
+        print(i, element)
+
+    # Results in
+    #   0 a
+    #   1 b
+    #   2 c
+    #   3 d
+    #   4 e
+
+
+* Here there is no need to actually index the list since the ``element`` variable already has the value ``another_list[i]``
+
+.. note::
+
+    Some people love using ``enumerate``, others feel it is unnecessary since they do not see indexing the list as an
+    imposition. Neither is more correct than the other --- use what you are comfortable with.
```

Comment:
Well. The argument is intent. The goal is to loop through the list, adding a layer of indirection through `range` is needlessly complicating. _Simplicity_ is the argument -- I'd argue that is indeed _more correct_.

Reply:
perhaps _neither is wrong_.

Reply:
The argument _against_ using it here, in intro CS, is avoiding another keyword that is particularly python-ie.

Reply:
It's a function, not a keyword. But maybe fair enough. Probably too early to introduce tuples and multiple returns.

I'd still maybe change the wording here. `enumerate` is more appropriate, but delves into some python features you have not (and will not?) touched on in the course.

Reply:
Oh, how about, yeet the "note" in question?

Reply:
Eh. Also fine.